.NET Aspire release 9.3 content (#3926)
dotnet/docs-aspire: docs/whats-new/dotnet-aspire-9.4.md

        return resourceBuilder;
    }
}
```

This custom deployment logic executes as follows from the `aspire deploy` command.

![aspire-deploy-whats-new](https://github.com/user-attachments/assets/15c6730d-8154-496a-be70-c67257ce5523)

Now, integration owners can create sophisticated `aspire deploy` workflows. This work also provides a foundation for advanced deployment automation scenarios.

> [!NOTE]
> While the <xref:Aspire.Hosting.Publishing.DeployingCallbackAnnotation> API is available in .NET Aspire 9.4, there are currently no built-in resources that natively support deployment callbacks. Built-in resource support for deployment callbacks will be added in the next version of .NET Aspire.

> [!IMPORTANT]
> 🧪 **Feature Flag**: The `aspire deploy` command is behind a feature flag and **disabled by default** in this release. It must be explicitly enabled for use with `aspire config set features.deployCommandEnabled true`

### 📃 Enhanced publish and deploy output

.NET Aspire 9.4 significantly improves the feedback and progress reporting during publish and deploy operations, providing clearer visibility into what's happening during deployment processes.

**Key improvements:**

> - **Enhanced progress reporting** with detailed step-by-step feedback during publishing
> - **Container runtime health checks** that provide clear status updates during container operations
> - **Better error messaging** with more descriptive information when deployments fail
> - **Improved publishing context** that tracks and reports on resource deployment status
> - **Cleaner output formatting** that makes it easier to follow deployment progress

Reviewer: ```suggestion
- **Enhanced progress reporting** with detailed step-by-step feedback during publishing
- **Cleaner output formatting** that makes it easier to follow deployment progress
- **Better error messaging** with more descriptive information when deployments fail
- **Improved publishing context** that tracks and reports on resource deployment status
- **Container build logs** provide clear status updates during container operations
```

### 👁️ Show/hide hidden resources in dashboard

.NET Aspire 9.4 introduces the ability to show or hide hidden resources in the dashboard, giving you complete visibility into your application's infrastructure components and internal resources that are normally hidden from view.

:::image type="content" source="media/dashboard-hidden-resources.png" lightbox="media/dashboard-hidden-resources.png" alt-text="Dashboard resources page with the show/hide hidden resources UI visible.":::

If there are no hidden resources in your Aspire app then the show/hide UI is disabled.

### 🏗️ Enhanced dashboard infrastructure with proxied endpoints

.NET Aspire 9.4 introduces significant infrastructure improvements to the dashboard system, implementing proxied endpoints that make dashboard launching more reliable and avoiding port reuse problems. This architectural enhancement resolves issues with dashboard connectivity during application startup and shutdown scenarios. The UI when the dashboard is attempting to reconnect has also been updated to be more reliable and with a new cohesive look and animation.

> 
> **Key improvements:**
> 
> - **Proxied dashboard endpoints** modeled as first-class proxied resources in DCP (Developer Control Plane)
> - **Automatic startup retry resilience** during DCP proxy startup eliminates connection failures from unclean dashboard shutdowns
> - **Port reuse problem resolution** for when dashboard ports weren't properly released from previous runs
> - **Improved reliability** in cases where the dashboard wasn't cleanly shut down from a previous run

Reviewer 2: ```suggestion
```

## ☁️ Azure goodies

### 🏷️ Consistent resource name exposure

.NET Aspire 9.4 now consistently exposes the actual names of deployed Azure resources through the <xref:Aspire.Hosting.Azure.NameOutputReference/> property. This enables applications to access the real Azure resource names that get generated during deployment, which is essential for scenarios requiring direct Azure resource coordination.

```csharp
var builder = DistributedApplication.CreateBuilder(args);

var storage = builder.AddAzureStorage("appstorage");
var signalr = builder.AddAzureSignalR("notifications");

// Access the actual deployed Azure resource names
var api = builder.AddProject<Projects.Api>("api")
                .WithEnvironment("STORAGE_NAME", storage.Resource.NameOutputReference)
                .WithEnvironment("SIGNALR_NAME", signalr.Resource.NameOutputReference);

builder.Build().Run();
```

This is particularly valuable for **external automation scripts** and **monitoring and alerting systems** that reference resources by their actual names. The `NameOutputReference` property is now available on all Azure resources.

> 

Reviewer 2: I shrunk it down!

var connectionString = secrets.AddSecret("connection-string", connectionStringParam);

// Add a secret with custom secret name in Key Vault
var apiKeyParam = builder.AddParameter("api-key", secret: true);
var apiKey = secrets.AddSecret("api-key", "ApiKey", apiKeyParam);

// Get a secret reference for consumption (for existing secrets)
var existingSecret = secrets.GetSecret("ExistingSecret");

// Use in your services
var webApi = builder.AddProject<Projects.WebAPI>("webapi")
    .WithEnvironment("CONNECTION_STRING", connectionString)
    .WithEnvironment("API_KEY", apiKey)
    .WithEnvironment("EXISTING_SECRET", existingSecret);
```

**Key features**:

- <xref.Aspire.Hosting.Azure.KeyVault.AzureKeyVaultResourceExtensions.AddSecret*/> method for adding new secrets to Key Vault from parameters or expressions
- <xref.Aspire.Hosting.Azure.KeyVault.AzureKeyVaultResourceExtensions.GetSecret*/> method for referencing existing secrets in Key Vault
- **Strongly-typed secret references** that can be used with `WithEnvironment()` for environment variables
- **Custom secret naming** support with optional `secretName` parameter

> ### 📀 Azure Storage client improvements
> 
> .NET Aspire 9.4 standardizes [Azure Storage client registration](../storage/azure-storage-integrations.md) with consistent naming conventions and enhanced keyed service support:
> 
> ```csharp
> var builder = WebApplication.CreateBuilder(args);
> 
> // Azure Table Storage with new standardized naming
> builder.AddAzureTableServiceClient("tables");
> builder.AddKeyedAzureTableServiceClient("primary-tables");
> 
> // Azure Queue Storage with consistent API
> builder.AddAzureQueueServiceClient("queues");
> builder.AddKeyedAzureQueueServiceClient("primary-queues");
> 
> // Keyed Blob Storage (enhanced support)
> builder.AddKeyedAzureBlobServiceClient("primary-blobs");
> 
> var app = builder.Build();
> ```
> 
> The new client registration methods provide consistent naming across all Azure Storage services and full support for keyed dependency injection scenarios.
> 
> #### 🔄 Flexible Azure Storage queue management
> 
> Working with [Azure Storage queues](../storage/azure-storage-queues-integration.md) often requires choosing between coarse-grained connection to entire storage accounts or manually managing individual queue configurations. .NET Aspire 9.4 introduces fine-grained queue modeling that lets you work with specific queues while maintaining clear separation of concerns.
> 
> ```csharp
> var builder = DistributedApplication.CreateBuilder(args);
> 
> var storage = builder.AddAzureStorage("storage");
> 
> // Model individual queues as first-class resources
> var orderQueue = storage.AddQueue("orders", "order-processing");
> var notificationQueue = storage.AddQueue("notifications", "user-notifications");
> 
> // Services get scoped access to specific queues
> builder.AddProject<Projects.OrderProcessor>("order-processor")
>        .WithReference(orderQueue);  // Only has access to order-processing queue
> 
> builder.AddProject<Projects.NotificationService>("notifications")
>        .WithReference(notificationQueue);  // Only has access to user-notifications queue
> 
> builder.Build().Run();
> ```
> 
> This approach provides better security isolation, clearer dependency modeling, and simplifies service configuration by injecting pre-configured `QueueClient` instances.

Reviewer: ```suggestion
#### 📥Resource Deep Linking for Azure Storage Queues

.NET Aspire 9.4 expands resource deep linking to include Azure Queue Storage queues, building on the model already used for Azure Blob Storage, Cosmos DB, etc.

You can now model individual storage queues directly in your app host, then inject scoped QueueClient instances into your services—making it easy to interact with queues without manually configuring connection strings or access.

**AppHost:**

```csharp
var builder = DistributedApplication.CreateBuilder(args);

var storage = builder.AddAzureStorage("storage");

// Model individual queues as first-class resources
var orderQueue = storage.AddQueue("orders", "order-processing");
var notificationQueue = storage.AddQueue("notifications", "user-notifications");

// Services get scoped access to specific queues
builder.AddProject<Projects.OrderProcessor>("order-processor")
       .WithReference(orderQueue);  // Only has access to order-processing queue

builder.AddProject<Projects.NotificationService>("notifications")
       .WithReference(notificationQueue);  // Only has access to user-notifications queue

builder.Build().Run();
```

**In the OrderProcessor project:**

```csharp
using Azure.Storage.Queues;

var builder = WebApplication.CreateBuilder(args);

// Register the queue client
builder.AddAzureQueue("orders");

var app = builder.Build();

// Minimal POST endpoint for image upload
app.MapPost("/process-order", async (QueueClient ordersQueue) =>
{
    // read a message for the queue
    var message = await ordersQueue.ReceiveMessageAsync();
    ProcessMessage(message);

    return Results.Ok();
});

app.Run();
```

This approach provides clean separation of concerns, secure container scoping, and minimal ceremony—ideal for microservices that interact with specific storage queues.
```

    .WithRealmImport("./realm.json", isReadOnly: false);  // Confusing parameter

// ✅ After (recommended):
var keycloak = builder.AddKeycloak("keycloak")
    .WithRealmImport("./realm.json");  // Clean, simple API

// If you need explicit read-only control:
var keycloak = builder.AddKeycloak("keycloak")
    .WithRealmImport("./realm.json", isReadOnly: true);  // Still available as overload
```

**Migration impact**: Remove the `isReadOnly` parameter from single-parameter `WithRealmImport()` calls - the method now defaults to appropriate behavior. Use the two-parameter overload if explicit control is needed.

### 🔧 Milvus configuration method updates

Milvus configuration has been updated with more descriptive method names:

```csharp
// ❌ Before (deprecated):
var milvus = builder.AddMilvus("milvus")
    .WithConfigurationBindMount("./milvus.yaml");  // Old method name

// ✅ After (recommended):
var milvus = builder.AddMilvus("milvus")
    .WithConfigurationFile("./milvus.yaml");  // Method renamed for clarity
```

**Migration impact**: Update method calls to use `WithConfigurationFile` instead of `WithConfigurationBindMount` for Milvus configuration.

> ### 🔒 Azure Cosmos DB connection string property changes
> 
> Azure Cosmos DB connection string handling has been updated to use the new connection string output approach:
> 
> ```csharp
> // ❌ Before (obsolete property):
> var cosmos = builder.AddAzureCosmosDB("cosmos");
> // cosmos.ConnectionString property is obsolete - use ConnectionStringExpression instead
> 
> // ✅ After (recommended):
> var cosmos = builder.AddAzureCosmosDB("cosmos");
> var connectionString = cosmos.ConnectionStringExpression;  // Use expression-based approach
> ```
> 
> **Migration impact**: Replace direct `ConnectionString` property access with `ConnectionStringExpression` for Azure Cosmos DB resources.
> 

Reviewer: ```suggestion
```

I don't think this Obsoletion is worth calling out. An extremely small number of customers would even be doing this. Plus also, there is no `cosmos.ConnectionString` or `cosmos.ConnectionStringExpression` properties. It needs to `.Resource` in the middle.

    })
    .WithEnvironment("CONNECTION_STRING", async (context, cancellationToken) =>
    {
        return await connectionString.Resource.GetValueAsync(cancellationToken);
    });

// For non-async contexts where blocking is acceptable:
var syncValue = apiKey.Resource.Value;  // Still works but may block
```

**Migration impact**: When working with `ParameterResource` values in async contexts, use the new `GetValueAsync()` method instead of the `Value` property to avoid potential deadlocks. For `WithEnvironment()` calls, prefer passing the parameter resource directly rather than accessing `.Value` synchronously.

### 🔧 Kafka configuration method changes

Kafka configuration has been updated with more descriptive method names:

```csharp
// ❌ Before (deprecated):
var kafka = builder.AddKafka("kafka")
    .WithConfigurationFile("./kafka.properties");  // Old method name

// ✅ After (recommended):
var kafka = builder.AddKafka("kafka")
    .WithConfigurationFile("./kafka.properties");  // Method renamed for clarity
```

**Migration impact**: Update method calls to use the new naming convention if you were using preview Kafka configuration methods.

> 

Reviewer 2: FowlerGPT artifact... good catch lol
